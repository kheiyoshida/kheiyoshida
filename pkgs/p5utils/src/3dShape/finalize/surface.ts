import p5 from 'p5'
import { Position3D } from '../../3d/types'
import { calcAverageVector, calcPerpendicularVector, isInTheSameSide } from '../tools'
import { ShapeVertex } from '../types'

type NormalDirectionOption = 'opposite' | 'same'

export const finalizeSurface = (
  surfaceVertices: ShapeVertex[],
  theOtherVertex: ShapeVertex,
  dir: NormalDirectionOption = 'opposite'
): void => {
  p.beginShape()
  p.normal(calcNormal(surfaceVertices, theOtherVertex, dir))
  surfaceVertices.forEach(vectorVertex)
  p.endShape()
}

/**
 * note: normal direction is on the opposite of the other vertex against surface by default
 *           |
 * normal <- | *
 *           |
 */
export const calcNormal = (
  surfaceVertices: ShapeVertex[],
  theOtherVertex: ShapeVertex,
  dir: NormalDirectionOption = 'opposite'
): p5.Vector => {
  const normal = calcPerpendicularVector(surfaceVertices)
  const surfaceCenter = calcAverageVector(surfaceVertices)
  const surfaceToTheOther = theOtherVertex.copy().sub(surfaceCenter)
  const makeOpposite = dir === 'opposite'
  if (isInTheSameSide(normal, surfaceToTheOther) === makeOpposite) {
    normal.mult(-1)
  }
  return normal
}

const vectorVertex = (v: ShapeVertex, i: number): void => {
  p.vertex(...(v.array() as Position3D), ...uv[i])
}

const uv = [
  [0, 0],
  [1, 0],
  [0.5, 1],
]
