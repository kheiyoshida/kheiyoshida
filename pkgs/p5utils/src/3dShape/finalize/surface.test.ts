import p5 from 'p5'
import { calcNormal, finalizeSurface } from './surface'

test(`${finalizeSurface.name}`, () => {
  const surface = [new p5.Vector(100, 0, 0), new p5.Vector(-100, 0, 0), new p5.Vector(0, 0, -100)]
  const theOther = new p5.Vector(0, -100, -50)
  const spyVertex = jest.spyOn(p, 'vertex')
  jest.spyOn(p, 'normal').mockImplementation()
  finalizeSurface(surface, theOther)
  surface.forEach((v) => {
    expect(spyVertex).toHaveBeenCalledWith(...v.array(), expect.any(Number), expect.any(Number))
  })
})

test.each([
  [
    [
      [100, 0, 0],
      [-100, 0, 0],
      [0, 0, -100],
    ],
    [0, -100, -50],
  ],
  [
    [
      [100, 0, 0],
      [-100, 0, 0],
      [0, 0, -100],
    ],
    [0, 100, -50],
  ],
])(`${calcNormal.name}`, (surface, theOther) => {
  const surfaceVertices = surface.map((s) => new p5.Vector(...s))
  const theOtherVertex = new p5.Vector(...theOther)
  const normal = calcNormal(surfaceVertices, theOtherVertex)
  expect(normal.array()).toMatchSnapshot()
})
