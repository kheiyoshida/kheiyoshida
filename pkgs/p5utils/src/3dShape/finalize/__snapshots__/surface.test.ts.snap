// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calcNormal 1`] = `
[
  0,
  1,
  0,
]
`;

exports[`calcNormal 2`] = `
[
  0,
  -1,
  0,
]
`;
